### ISSUE TITLE
BioTokenizerPostProcessor.revertAggressiveTokenization introduces misaligned offsets.

### ISSUE THREAD

**Issue by myedibleenso** [clulab/processors#240]
@MihaiSurdeanu, [`org.clulab.processors.clu.bio.BioTokenizerPostProcessor.revertAggressiveTokenization`](https://github.com/clulab/processors/blob/901a0dd459ef5053e70a412f943bbd71dabf3ce2/main/src/main/scala/org/clulab/processors/clu/bio/BioTokenizerPostProcessor.scala#L231-L265) sometimes introduces misaligned offsets.

I've isolated this issue by commenting and uncommenting some of the calls in [`BioTokenizerPostProcessor.BioTokenizerPostProcessor.process`](https://github.com/clulab/processors/blob/901a0dd459ef5053e70a412f943bbd71dabf3ce2/main/src/main/scala/org/clulab/processors/clu/bio/BioTokenizerPostProcessor.scala#L28-L59).

#### MWE
Consider this sentence:

>We describe here a novel molecule of this group termed SIT (SHP2 interacting transmembrane adaptor protein).

In the following example:

```scala
import org.clulab.processors.clu.BioCluProcessor

val proc = new BioCluProcessor()
val text = "We describe here a novel molecule of this group termed SIT (SHP2 interacting transmembrane adaptor protein)."
val doc.mkDocument(text)
text.endOffsets
```

#### Before `.revertAgressiveTokenization`:
`Array(2, 11, 16, 18, 24, 33, 36, 41, 47, 54, 58, 60, 64, 76, 90, 98, 106, 107, 108)`

#### After `.revertAgressiveTokenization`:
`Array(2, 11, 16, 18, 24, 33, 36, 41, 47, 54, 58, 64, 64, 76, 90, 98, 106, 111, 108)`

If you compare the penultimate and final end offsets, you'll notice that there is a non-final token ending after the token which follows it.

**Comment by MihaiSurdeanu** [comment]
This sounds like my bug. I'll take a look tonight.

**Comment by myedibleenso** [comment]
I haven't tried with `BioNLPProcessor`.  Have you tried with `BioCluProcessor`?

**Comment by MihaiSurdeanu** [comment]
Ok, I can replicate the bug with BioClu. I'll look into it.
But, for the paper, you should use BioNLPProc. This has been more extensively tested.

**Comment by MihaiSurdeanu** [comment]
This happens because the ")" is represented as "-RRB-". revertAggressiveTokenization keeps the start positions from the original tokens (")"), but sets the end positions based on the length of the token, hence the weird end position.

We should decide the type of behavior we want: should we keep the start/end positions from the original tokens, regardless of the value stored in the token?

**Comment by myedibleenso** [comment]
On a related note, I noticed that these two processors seem to have different defaults for `kbsWithTokensWithValidSlashes`:
- https://github.com/clulab/processors/blob/70db37eac25d813bc092e0e5eede0d5773b63cd5/corenlp/src/main/scala/org/clulab/processors/bionlp/BioNLPProcessor.scala#L33
   - https://github.com/clulab/processors/blob/901a0dd459ef5053e70a412f943bbd71dabf3ce2/corenlp/src/main/scala/org/clulab/processors/bionlp/ner/KBLoader.scala#L43
   - https://github.com/clulab/processors/blob/901a0dd459ef5053e70a412f943bbd71dabf3ce2/corenlp/src/main/resources/reference.conf#L81-L92
- https://github.com/clulab/processors/blob/901a0dd459ef5053e70a412f943bbd71dabf3ce2/main/src/main/resources/cluprocessorbio.conf#L11-L18

It looks like `BioNLPProcessor` also makes use of `Phase-3-Override-tsv.gz`.  I don't think this is the cause of the misalignment, but should we change the defaults to be consistent?

**Comment by myedibleenso** [comment]
Whoops!  Nevermind.  I just noticed that particular line is actually commented out...

**Comment by myedibleenso** [comment]
I believe the offsets are only useful if they correspond to the original text.  If they don't, why bother to generate them at all, right?

**Comment by MihaiSurdeanu** [comment]
I think you're right.

**Comment by myedibleenso** [comment]
Hopefully I didn't misinterpret your comment :)

Here is how I handled the length issue for better correspondence with the BioNLP ST data: 
- https://github.com/clulab/processors/blob/436390d1ef530b1a41c66f479a516ff804cb3fb9/main/src/main/scala/org/clulab/processors/clu/bio/BioTokenizerPostProcessor.scala#L31-L35

I think that may be incomplete, though, as it only covers bracket-like stuff.

**Comment by MihaiSurdeanu** [comment]
I think that is fine for this dataset.
In Reach, another change that affects length is changing unicode greek
letters to their ascii representation. I don't think this appears in BioNLP
however.

**Comment by MihaiSurdeanu** [comment]
I think it's Ok to stay with this fix for now. I would like to reorganize this code anyway, to preserve the original raw tokens, which would solve this problem.

**Comment by MihaiSurdeanu** [comment]
This is fixed by the latest master (7.4.0-SNAPSHOT).
Closing for now.
